src/compiler/c.rs

```diff
@@ -445,6 +557,379 @@ where
+            return Ok(false);
+        } else if slice.starts_with(b"___________") && (start == 0 || bytes[start - 1] == b'\n') {
+            // Unfortunately the distcc-pump wrapper outputs standard output lines:
+            // __________Using distcc-pump from /usr/bin
+            // __________Using # distcc servers in pump mode
+            // __________Shutting down distcc-pump include server
+            digest.update(&bytes[hash_start..start]);
+            while start < total_len && slice[0] != b'\n' {
+                start += 1;
+                if start < total_len {
+                    slice = &bytes[start..];
+                }
+            }
+            slice = &bytes[start..];
+            if slice[0] == b'\n' {
+                start += 1;
+            }
+            hash_start = start;
+            continue;
+        } else {
+            start += 1;
+        }
+    }
+    digest.update(&bytes[hash_start..]);
+
+    Ok(true)
+}
+
+/// Copied from cargo.
+///
+/// Normalize a path, removing things like `.` and `..`.
+///
+/// CAUTION: This does not resolve symlinks (unlike
+/// [`std::fs::canonicalize`]).
+pub fn normalize_path(path: &Path) -> PathBuf {
+    use std::path::Component;
+    let mut components = path.components().peekable();
+    let mut ret = if let Some(c @ Component::Prefix(..)) = components.peek().cloned() {
+        components.next();
+        PathBuf::from(c.as_os_str())
+    } else {
+        PathBuf::new()
+    };
+
+    for component in components {
+        match component {
+            Component::Prefix(..) => unreachable!(),
+            Component::RootDir => {
+                ret.push(component.as_os_str());
+            }
+            Component::CurDir => {}
+            Component::ParentDir => {
+                ret.pop();
+            }
+            Component::Normal(c) => {
+                ret.push(c);
+            }
+        }
+    }
+    ret
+}
+
+// Returns false if the include file was "too new" and therefore should disable
+// the direct mode, otherwise true.
+#[allow(clippy::too_many_arguments)]
+fn remember_include_file(
+    mut path: &[u8],
+    input_file: &Path,
+    cwd: &Path,
+    included_files: &mut HashMap<PathBuf, String>,
+    digest: &mut Digest,
+    system: bool,
+    config: DirectModeConfig,
+    time_of_compilation: std::time::SystemTime,
+) -> Result<bool> {
+    // TODO if precompiled header.
+    if path.len() >= 2 && path[0] == b'<' && path[path.len() - 1] == b'>' {
+        // Typically <built-in> or <command-line>.
+        digest.update(path);
+        return Ok(true);
+    }
+
+    if system && config.skip_system_headers {
+        // Don't remember this system header, only hash its path.
+        digest.update(path);
+        return Ok(true);
+    }
+
+    let original_path = path;
+    // Canonicalize path for comparison; Clang uses ./header.h.
+    if path.starts_with(b"./") {
```

Comment:
It appears that in some compiler versions and platforms, `\` can be valid. `ccache` does some blanket normalization of backslashes to slashes, but we work on native `Path[Buf]` structs. Long story short, I can just add `|| path.starts_with(br".\")` here, since it's the only place we directly compare bytes, aside from the `<` and `>` check right above.